File: csharp/item/ItemFunctions.cs
using System;
using System.Collections.Generic;
using com.ultracart.admin.v2.Api;
using com.ultracart.admin.v2.Model;

namespace SdkSample.item
{
    public class ItemFunctions
    {
        /// <summary>
        /// Creates a sample item with random ID
        /// </summary>
        /// <returns>The newly created item id</returns>
        public static string InsertSampleItem()
        {
            Random random = new Random();
            string chars = "ABCDEFGH";
            char[] stringChars = new char[chars.Length];

            for (int i = 0; i < stringChars.Length; i++)
            {
                stringChars[i] = chars[random.Next(chars.Length)];
            }

            string rand = new string(stringChars);
            string itemId = $"sample_{rand}";

            Console.WriteLine($"InsertSampleItem will attempt to create item {itemId}");
            ItemApi itemApi = Samples.GetItemApi();

            Item newItem = new Item();
            newItem.MerchantItemId = itemId;

            ItemPricing pricing = new ItemPricing();
            pricing.Cost = 9.99m;
            newItem.Pricing = pricing;

            newItem.Description = $"Sample description for item {itemId}";

            ItemContentMultimedia multimedia = new ItemContentMultimedia();
            multimedia.Url = "https://upload.wikimedia.org/wikipedia/en/7/73/Mr._Clean_logo.png";
            multimedia.Code = "default"; // <-- use 'default' to make this the default item.
            multimedia.Description = "Some random image i nabbed from wikipedia";

            ItemContent content = new ItemContent();
            content.Multimedia = new List<ItemContentMultimedia> { multimedia }; // <- notice this is an array
            newItem.Content = content;

            string expand = "content.multimedia"; // I want to see the multimedia returned on the newly created object.

            Console.WriteLine("InsertItem request object follows:");
            Console.WriteLine(newItem);

            ItemResponse apiResponse = itemApi.InsertItem(newItem, expand);

            Console.WriteLine("InsertItem response object follows:");
            Console.WriteLine(apiResponse);

            return apiResponse.Item.MerchantItemId;
        }

        
        /// <summary>
        /// Creates a sample item with random ID
        /// </summary>
        /// <returns>The newly created merchant item oid</returns>
        public static int InsertSampleItemAndGetOid()
        {
            Random random = new Random();
            string chars = "ABCDEFGH";
            char[] stringChars = new char[chars.Length];

            for (int i = 0; i < stringChars.Length; i++)
            {
                stringChars[i] = chars[random.Next(chars.Length)];
            }

            string rand = new string(stringChars);
            string itemId = $"sample_{rand}";

            Console.WriteLine($"InsertSampleItem will attempt to create item {itemId}");
            ItemApi itemApi = Samples.GetItemApi();

            Item newItem = new Item();
            newItem.MerchantItemId = itemId;

            ItemPricing pricing = new ItemPricing();
            pricing.Cost = 9.99m;
            newItem.Pricing = pricing;

            newItem.Description = $"Sample description for item {itemId}";

            ItemContentMultimedia multimedia = new ItemContentMultimedia();
            multimedia.Url = "https://upload.wikimedia.org/wikipedia/en/7/73/Mr._Clean_logo.png";
            multimedia.Code = "default"; // <-- use 'default' to make this the default item.
            multimedia.Description = "Some random image i nabbed from wikipedia";

            ItemContent content = new ItemContent();
            content.Multimedia = new List<ItemContentMultimedia> { multimedia }; // <- notice this is an array
            newItem.Content = content;

            string expand = "content.multimedia"; // I want to see the multimedia returned on the newly created object.

            Console.WriteLine("InsertItem request object follows:");
            Console.WriteLine(newItem);

            ItemResponse apiResponse = itemApi.InsertItem(newItem, expand);

            Console.WriteLine("InsertItem response object follows:");
            Console.WriteLine(apiResponse);

            return apiResponse.Item.MerchantItemOid;
        }

        
        /// <summary>
        /// Deletes a sample item by its oid
        /// </summary>
        /// <param name="merchantItemOid">The item oid to be deleted</param>
        public static void DeleteSampleItemByOid(int merchantItemOid)
        {
            ItemApi itemApi = Samples.GetItemApi();
            Console.WriteLine($"Calling DeleteItem({merchantItemOid})");
            itemApi.DeleteItem(merchantItemOid);
        }
        
        
        /// <summary>
        /// Deletes a sample item by its oid
        /// </summary>
        /// <param name="merchantItemOid">The item id to be deleted</param>
        public static void DeleteSampleItem(string merchantItemId)
        {
            ItemApi itemApi = Samples.GetItemApi();
            int merchantItemOid = itemApi.GetItemByMerchantItemId(merchantItemId).Item.MerchantItemOid;

            Console.WriteLine($"Calling DeleteItem({merchantItemOid})");
            itemApi.DeleteItem(merchantItemOid);
        }

        /// <summary>
        /// Creates a sample digital item
        /// </summary>
        /// <param name="externalId">Optional external ID for the digital item</param>
        /// <returns>The digital item oid for the newly created item</returns>
        public static int InsertSampleDigitalItem(string externalId = null)
        {
            string imageUrl =
                "https://upload.wikimedia.org/wikipedia/commons/b/b7/Earth_%2816530938850%29.jpg"; // picture of the earth

            ItemDigitalItem digitalItem = new ItemDigitalItem();
            digitalItem.ImportFromUrl = imageUrl;
            digitalItem.Description = "The Earth";
            digitalItem.ClickWrapAgreement = "By purchasing this item, you agree that it is Earth";

            if (externalId != null)
            {
                digitalItem.ExternalId = externalId;
            }

            Console.WriteLine("InsertDigitalItem request object follows:");
            Console.WriteLine(digitalItem);

            ItemApi itemApi = Samples.GetItemApi();
            ItemDigitalItemResponse apiResponse = itemApi.InsertDigitalItem(digitalItem);

            Console.WriteLine("InsertDigitalItem response object follows:");
            Console.WriteLine(apiResponse);

            return apiResponse.DigitalItem.DigitalItemOid;
        }

        /// <summary>
        /// Deletes a sample digital item by its oid
        /// </summary>
        /// <param name="digitalItemOid">The primary key of the digital item to be deleted</param>
        public static void DeleteSampleDigitalItem(int digitalItemOid)
        {
            ItemApi itemApi = Samples.GetItemApi();

            Console.WriteLine($"Calling DeleteDigitalItem({digitalItemOid})");
            itemApi.DeleteDigitalItem(digitalItemOid);
        }
    }
}

File: typescript/item/ItemFunctions.ts
// noinspection DuplicatedCode

import { itemApi } from '../api';
import { Item, ItemContent, ItemContentMultimedia, ItemPricing, ItemResponse, ItemDigitalItem, ItemDigitalItemResponse } from 'ultracart_rest_api_v2_typescript';

export class ItemFunctions {
    /// <summary>
    /// Creates a sample item with random ID
    /// </summary>
    /// <returns>The newly created item id</returns>
    public static async insertSampleItem(): Promise<string> {
        const chars = "ABCDEFGH";
        const rand = Array.from({ length: chars.length }, () => chars[Math.floor(Math.random() * chars.length)]).join('');
        const itemId = `sample_${rand}`;

        console.log(`InsertSampleItem will attempt to create item ${itemId}`);

        const newItem: Item = {
            merchant_item_id: itemId,
            pricing: {
                cost: 9.99
            } as ItemPricing,
            description: `Sample description for item ${itemId}`,
            content: {
                multimedia: [{
                    url: "https://upload.wikimedia.org/wikipedia/en/7/73/Mr._Clean_logo.png",
                    code: "default", // <-- use 'default' to make this the default item.
                    description: "Some random image i nabbed from wikipedia"
                } as ItemContentMultimedia]
            } as ItemContent
        };

        const expand = "content.multimedia"; // I want to see the multimedia returned on the newly created object.

        console.log("InsertItem request object follows:");
        console.log(newItem);

        const apiResponse: ItemResponse = await itemApi.insertItem({ item: newItem, expand: expand });

        console.log("InsertItem response object follows:");
        console.log(apiResponse);

        return apiResponse.item === undefined || apiResponse.item.merchant_item_id === undefined  ? "ERR_NO_ID" : apiResponse.item.merchant_item_id;
    }

    /// <summary>
    /// Creates a sample item with random ID
    /// </summary>
    /// <returns>The newly created merchant item oid</returns>
    public static async insertSampleItemAndGetOid(): Promise<number> {
        const chars = "ABCDEFGH";
        const rand = Array.from({ length: chars.length }, () => chars[Math.floor(Math.random() * chars.length)]).join('');
        const itemId = `sample_${rand}`;

        console.log(`InsertSampleItem will attempt to create item ${itemId}`);

        const newItem: Item = {
            merchant_item_id: itemId,
            pricing: {
                cost: 9.99
            } as ItemPricing,
            description: `Sample description for item ${itemId}`,
            content: {
                multimedia: [{
                    url: "https://upload.wikimedia.org/wikipedia/en/7/73/Mr._Clean_logo.png",
                    code: "default", // <-- use 'default' to make this the default item.
                    description: "Some random image i nabbed from wikipedia"
                } as ItemContentMultimedia]
            } as ItemContent
        };

        const expand = "content.multimedia"; // I want to see the multimedia returned on the newly created object.

        console.log("InsertItem request object follows:");
        console.log(newItem);

        const apiResponse: ItemResponse = await itemApi.insertItem({ item: newItem, expand: expand });

        console.log("InsertItem response object follows:");
        console.log(apiResponse);

        return apiResponse.item === undefined || apiResponse.item.merchant_item_oid === undefined  ? 0 : apiResponse.item.merchant_item_oid;
    }

    /// <summary>
    /// Deletes a sample item by its oid
    /// </summary>
    /// <param name="merchantItemOid">The item oid to be deleted</param>
    public static async deleteSampleItemByOid(merchantItemOid: number): Promise<void> {
        console.log(`Calling DeleteItem(${merchantItemOid})`);
        await itemApi.deleteItem({ merchantItemOid: merchantItemOid });
    }

    /// <summary>
    /// Deletes a sample item by its id
    /// </summary>
    /// <param name="merchantItemId">The item id to be deleted</param>
    public static async deleteSampleItem(merchantItemId: string): Promise<void> {
        const merchantItemOid: number|undefined = (await itemApi.getItemByMerchantItemId({ merchantItemId: merchantItemId })).item?.merchant_item_oid;

        console.log(`Calling DeleteItem(${merchantItemOid})`);
        if(merchantItemOid !== undefined) {
            await itemApi.deleteItem({ merchantItemOid: merchantItemOid });
        }
    }

    /// <summary>
    /// Creates a sample digital item
    /// </summary>
    /// <param name="externalId">Optional external ID for the digital item</param>
    /// <returns>The digital item oid for the newly created item</returns>
    public static async insertSampleDigitalItem(externalId?: string): Promise<number> {
        const imageUrl = "https://upload.wikimedia.org/wikipedia/commons/b/b7/Earth_%2816530938850%29.jpg"; // picture of the earth

        const digitalItem: ItemDigitalItem = {
            import_from_url: imageUrl,
            description: "The Earth",
            click_wrap_agreement: "By purchasing this item, you agree that it is Earth"
        };

        if (externalId !== undefined) {
            digitalItem.external_id = externalId;
        }

        console.log("InsertDigitalItem request object follows:");
        console.log(digitalItem);

        const apiResponse: ItemDigitalItemResponse = await itemApi.insertDigitalItem({ digitalItem: digitalItem });

        console.log("InsertDigitalItem response object follows:");
        console.log(apiResponse);

        return apiResponse.digital_item === undefined || apiResponse.digital_item.digital_item_oid === undefined  ? 0 : apiResponse.digital_item.digital_item_oid;
    }

    /// <summary>
    /// Deletes a sample digital item by its oid
    /// </summary>
    /// <param name="digitalItemOid">The primary key of the digital item to be deleted</param>
    public static async deleteSampleDigitalItem(digitalItemOid: number): Promise<void> {
        console.log(`Calling DeleteDigitalItem(${digitalItemOid})`);
        await itemApi.deleteDigitalItem({ digitalItemOid: digitalItemOid });
    }
}